Makefile

```diff
@@ -0,0 +1,8 @@
+default:
+	mkdir -p build
+	echo "Building tools.."
+	$(MAKE) -C tools
+	echo "Building middleware.."
+	$(MAKE) -C middleware
```

Comment:
This simple `Makefile` suggests that it can just be run. But for the middleware, first `make envinit` is needed, so a simple `make` on the top level breaks without further explanation.

```
$ make
mkdir -p build
echo "Building tools.."
Building tools..
make -C tools
make[1]: Entering directory '/home/stadicus/github/stadicus/bitbox-base/tools'
make -C bbbfancontrol
make[2]: Entering directory '/home/stadicus/github/stadicus/bitbox-base/tools/bbbfancontrol'
cp bbbfancontrol.service /home/stadicus/github/stadicus/bitbox-base/tools/bbbfancontrol/../../build/
GOARCH=arm64 go build -o /home/stadicus/github/stadicus/bitbox-base/tools/bbbfancontrol/../../build/bbbfancontrol
make[2]: Leaving directory '/home/stadicus/github/stadicus/bitbox-base/tools/bbbfancontrol'
make[1]: Leaving directory '/home/stadicus/github/stadicus/bitbox-base/tools'
echo "Building middleware.."
Building middleware..
make -C middleware
make[1]: Entering directory '/home/stadicus/github/stadicus/bitbox-base/middleware'
GOARCH=arm64 go build -o /home/stadicus/github/stadicus/bitbox-base/middleware/../build/base-middleware ./src/
src/main.go:11:2: cannot find package "github.com/btcsuite/btcd/rpcclient" in any of:
	/usr/local/go/src/github.com/btcsuite/btcd/rpcclient (from $GOROOT)
	/home/stadicus/go/src/github.com/btcsuite/btcd/rpcclient (from $GOPATH)
src/main.go:13:2: cannot find package "github.com/fiatjaf/lightningd-gjson-rpc" in any of:
	/usr/local/go/src/github.com/fiatjaf/lightningd-gjson-rpc (from $GOROOT)
	/home/stadicus/go/src/github.com/fiatjaf/lightningd-gjson-rpc (from $GOPATH)
src/main.go:14:2: cannot find package "github.com/gorilla/mux" in any of:
	/usr/local/go/src/github.com/gorilla/mux (from $GOROOT)
	/home/stadicus/go/src/github.com/gorilla/mux (from $GOPATH)
src/main.go:15:2: cannot find package "github.com/gorilla/websocket" in any of:
	/usr/local/go/src/github.com/gorilla/websocket (from $GOROOT)
	/home/stadicus/go/src/github.com/gorilla/websocket (from $GOPATH)
Makefile:12: recipe for target 'aarch64' failed
make[1]: *** [aarch64] Error 1
make[1]: Leaving directory '/home/stadicus/github/stadicus/bitbox-base/middleware'
Makefile:2: recipe for target 'default' failed
make: *** [default] Error 2
```

Reply:
> This simple Makefile suggests that it can just be run.

That's certainly the intent!

> But for the middleware, first make envinit is needed, so a simple make on the top level breaks without further explanation.

Ah, I see what's going on. I have all my source code under `${GOPATH}/src` (due to working on other Go software), so for me it "Just Works".

We'd want to ideally support a user that has `go` installed but no `${GOPATH}` and/or has checked out the repository outside `${GOPATH}` too, though. At the very least, we should detect this case and give a friendly hint as to what to change.

I'll think about how to best achieve this and push another commit that addresses this flaw. Thanks for noticing it!

Reply:
Yes, I agree. I have set up Go and my $GOPATH is set, but as this is not a pure Go repository it's not checked out inside the GOPATH. That might be confusing, so we should decide how to handle that best. 

Do you think it's necessary to have everything, including Armbian and other scripts, inside the GOPATH?

Reply:
> Yes, I agree. I have set up Go and my $GOPATH is set, but as this is not a pure Go repository it's not checked out inside the GOPATH. That might be confusing, so we should decide how to handle that best.
>
> Do you think it's necessary to have everything, including Armbian and other scripts, inside the GOPATH?

We do need to assume that the repo is checked out under `${GOPATH}/src`, if we want the current `vendor/` system used to provide the dependencies for `middleware/` to be found by the Go toolchains. If you are interested in the details, see [this link](https://github.com/golang/go/wiki/GOPATH#vendor-directories-ignored-outside-of-gopath).

It's indeed a bit complicated / messy, and moreso since the Go ecosystem is also partially through a migration away from `GOPATH` (towards "Go modules") at time of writing..

I think what's easiest for now is to just add a simple check that the environment is in a good state (which in our context / current build setup means that a) `${GOPATH}` is set and b) the repo is checked out under `${GOPATH}/src`). That way, users who want to do the Go build will get clear instructions on what's going wrong if their environment is not ready for the build.

Next step might be either of:
1. dockerizing the Go builds, or otherwise providing a known good environment so user's who would like to build an Armbian image don't need the Go toolchain + configuration at all
1. switching to using "Go modules", which doesn't require any specific `${GOPATH}` structure, if we can assume that Go 1.11+ is available in the build environment (see [this link](https://github.com/golang/go/wiki/Modules#how-do-i-use-vendoring-with-modules-is-vendoring-going-away) if you are interested in the details)

There's a lot of details here, but hopefully the commit I've partially written will make it clear what's going.